source-mysql/replication.go

```diff
@@ -769,6 +769,36 @@ func (rs *mysqlReplicationStream) ActivateTable(ctx context.Context, streamID st
 				}
 			}
 		}
+
+		// This is a temporary piece of migration logic added in March of 2024. In the PR
+		// https://github.com/estuary/connectors/pull/1336 the ordering of enum cases in this
+		// metadata was changed to simplify the decoding logic. Specifically the order of the
+		// cases was changed from ["A", "B", ""] to ["", "A", "B"] to more directly mirror how
+		// MySQL represents the illegal-enum value "" as integer 0. However this introduced a bug
+		// when the new indexing code was used with older metadata from before that change, and
+		// by the time this was discovered there were captures in production with the new ordering
+		// in their checkpointed metadata, so a simple revert was not an option.
+		//
+		// The solution is to push forward and add the missing migration step, so that upon table
+		// activation any metadata containing the old ordering will be updated to the new. Once all
+		// metadata in production has been so updated, it will be safe to remove this logic and the
+		// associated 'TestEnumDecodingFix' test case.
+		for column, columnType := range metadata.Schema.ColumnTypes {
```

Comment:
What's the mode of action by which the updated `rs.tables.metadata[streamID]` gets persisted back into the connector state? Will it happen if the DB is otherwise idle? (I think I know the answers, just want to explicitly have the discussion)

Reply:
Tables which are already active (that is, we've previously started capturing them) at startup will run this logic before replication streaming starts. So while the main replication loop does only send metadata updates before blocking for the next replication event, these updates should be ready to go before that and thus take effect immediately even if the database is idle.

Also we generate a watermark change every couple minutes which would ensure we make progress even if the above wasn't the case.